README.md

```diff
@@ -1,19 +1,122 @@
+              ker1 bias1                        -- layer1 kernel, shape (c_out, c_in, kh+1, kw+1); and bias, shape (c_out)
+              ker2 bias2                        -- layer2 kernel, shape (c_out, c_out, kh+1, kw+1); and bias, shape (c_out)
+              weigthsDense                      -- dense weights shape (n_hidden, c_out, ((h+kh)/2 + kh)/2, ((w+kw)/2 + kw)/2) 
+              biasesDense                       -- dense biases shape (n_hidden)
+              weigthsReadout biasesReadout      -- readout layer weights (10, n_hidden) and bias (10)
+              =                                 --> output (10, batch_size)
+  do
+  t1 <- convMnistLayerS ker1 (constant x) bias1
+  t2 <- convMnistLayerS ker2 t1 bias2
+  let m1 = mapS reshapeS t2
+      m2 = from2S (transpose2 (fromS2 m1))  -- TODO: add permuation transposeS
+      denseLayer = weigthsDense <>$ m2 + asColumnS biasesDense
+  denseRelu <- reluAct denseLayer
+  returnLet $ weigthsReadout <>$ denseRelu + asColumnS biasesReadout
+```
+But we don't just want the shapes in comments, we want them in the type system:
+```Haskell
+convMnistTwoS
+  :: forall kh kw num_hidden c_out h w c_in batch_size d r m.
+     ( KnownNat kh, KnownNat kw, KnownNat num_hidden, KnownNat c_out, KnownNat h, KnownNat w, KnownNat batch_size
+     , c_in ~ 1 -- greyscale
+     , 1 <= kh  -- kh greater than 1
+     , 1 <= kw  -- kw greater than 1
+     , DualMonad d r m )
+  => OS.Array '[batch_size, c_in, h, w] r
+  -> DualNumber d (OS.Array '[ c_out, c_in , kh + 1, kw + 1 ] r)
+  -> DualNumber d (OS.Array '[c_out] r)
+  -> DualNumber d (OS.Array '[ c_out, c_out, kh + 1, kw + 1 ] r)
+  -> DualNumber d (OS.Array '[c_out] r)
+  -> DualNumber d (OS.Array '[ num_hidden, c_out,
+                                 GHC.TypeLits.* ((h + kh) `Div` 2 + kh) `Div` 2
+                                 GHC.TypeLits.* ((w + kw) `Div` 2 + kw) `Div` 2 -- [buglet? kh->kw? -awf]
+                             ] r)
+  -> DualNumber d (OS.Array '[num_hidden] r)
+  -> DualNumber d (OS.Array '[10, num_hidden] r)
+  -> DualNumber d (OS.Array '[10] r)
+  -> m (DualNumber d (OS.Array '[10, batch_size] r))
+```
+
+Several sample neural networks for MNIST digit classification are gathered in
 
 https://github.com/Mikolaj/horde-ad/tree/master/src/HordeAd/Tool
 
 Among them are recurrent, convolutional and fully connected networks based on fully typed tensors (sizes of all dimensions are tracked in the types) as well as weakly typed fully connected networks built with, respectively, matrices, vectors and raw scalars (most flexible but slowest).
 
 
+# Derivative checking
+
+Elsewhere in the same file is a computation of the forward derivative of the function and a QuickCheck test relating it to the gradient. It uses the same definition of the objective function and the same glue code for grouping parameters, etc. The ratio of signal to noise (maths to glue code) is much higher in more complex functions, e.g., neural networks. 
```

Comment:
This could be refined once we settle on the text above it, because by this point you've already shown a fragment of neural network definition, omitting the glue/noise. However, it's still very valuable IMHO that the QuickCheck tests exist for many small and big examples, including both examples in the README and many other mentioned in the README. And they are not smoke tests, but randomized checks of the fundamental formula you suggested

> dt <.> derivativeFromFun f ds = gradientFromFun dt f <.> ds

which the code was not designed to compute nor fine-tuned to pass. The tests pass as a happy coincidence, because the code is faithful to maths and they can be expressed as cleanly as above, because the API mimics maths.